design/simple-global-load-balancing.md

```diff
@@ -0,0 +1,938 @@
+
+It would be great if it was possible to use Metallb to create global load balancers.
+
+Global Load Balancers support the use case of deploying the sam applications in active/active mode across different datacenters.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+To support global load balancing for those scenario in which metallb and BGP is viable, without having to rely on external infrastructure.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Enhance Metallb with the ability to coordinate with other Metallb instances to create global load balancers
+- Allow tenants of a clusters to self-service global load balancers.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- support for advanced global load balancing strategies
+- support for BGP-router initiated health checks
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+It is simple to show that if two LoadBalancer Services in two different clusters are created with the same VIP and Metallb is configured with BGP, the traffic will be load balanced across the two clusters.
+
+One option would be to have the tenants requesting the LoadBalancer Services to input the same VIP across the multi0ple clusters that need to be globally load balanced.
+
+This would mean, though, that the tenants would have to have to be involved in the IPAM process for the VIPs, which is not desirable.
+
+That said at the moment there is no simple way to coordinate different instances of Metallb to pick the sam VIP for the same LoadBalancer service when deployed across different clusters.
+
+So some coordination is needed between the clusters across which one needs to establish global load balancing. And yet the metallb operator is cluster-bound and cannot "see" beyond the cluster it's running on.
+
+We propose to use DNS as a means of coordination between the various metallb instances.
+
+It should be possible to express the desire to have the VIP of a LoadBalancer service assigned by doing a dns lookup with a given FQDN.
+
+For example if a LoadBalancer service is annotated with: `metallb.io/assign-from-fqdn: myglobal.service.example.io`
```

Comment:
If I understand Metalb correclty, `IPaddressPool` is a highly powerful API which you wouldn't give access to to the developer. But the fqdn from which the IP needs to be takes is developer input. 
So I don;t think this is a good approach. The main requirement is that as a developer I should be able to define my globally loadbalanced services

Reply:
I agree that you can't give the IPAddressPool to the user, but it's for the same reason you can't give him the annotation: it will give the tenant of the cluster the ability to pick whatever ip he wants. Also, as you described, this must be agreed with the cluster admin in any case, as the ip will have to match one available in the IPAddressPool. In general, the available ips are under the admin's control.

So the right workflow has to be "I have this service, I want to tie it to this fqdn, please admin configure me to do that".
As I wrote, we need to move the responsability of the ip assignement away from the service, otherwise it'd be easy to abuse of the feature and steal ips to other tenants.

Reply:
The FQDN must be user input. Also we cannot introduce a process by which the user has to ask something to the platform engineer. It would be like if the user had to ask the platform engineering to create an Ingress after the chose the fqdn. The process has to be completely self-serviceable.
`but it's for the same reason you can't give him the annotation: it will give the tenant of the cluster the ability to pick whatever ip he wants`: how do you reach that conclusion. 
As a user I request the VIPs behind a given FQDN, the system will accept that request only after passing the checks described in the doc. Otherwise the system will reject the request. Exactly as it happens with Ingresses if a user requests an Ingress with an FQDN that is already claimed.

Reply:
> The FQDN must be user input. Also we cannot introduce a process by which the user has to ask something to the platform engineer. It would be like if the user had to ask the platform engineering to create an Ingress after the chose the fqdn. The process has to be completely self-serviceable. `but it's for the same reason you can't give him the annotation: it will give the tenant of the cluster the ability to pick whatever ip he wants`: how do you reach that conclusion. 

Because that fqdn can be associated to any ip (available, if we add the check), not only the one associated to the service. Additionally, if the user is asking for the IP associated to a given fqdn, then that IP must be registered (and kept in sync, some how), because the addresspools might go out of sync.

Having the IPAddressPool follow that IP in an automatic fashion removes part of the burden of having to change it when / if the dns entry changes. Once we have an IPAddressPool fed by the dns resolution, the user can ask to have that ip pinned to his service, or he can specify the pool annotation on the service to make sure the service is fed by that IP addresspool. This would also open the door to other ways to feed the ipaddresspools from different sources.

Reply:
I agree with the risk of the returned IP not fitting the associated address pool. And I addressed that risk in with the proposed checks in the caveat section.
To reiterate the concept, the proposed annotation is a hint to the IPAM part of the operator: given an AddressPool and an IP being requested, instead of the normal IPAM process, lookup the IP via DNS. After that, all the standard checks apply. Meaning if the IP is not in the CIDR range, or if it is already taken, the IP should not be assigned. In that case the Service should not receive an IP and an event should be emitted describing the problem.

Reply:
I agree that whatever the solution is here, it has to be self serviceable by the developers. any friction between developers owning their ideation to production workflow including full CI, continuous delivery, and continuous deployment, along with all associated deployment config, in this case including global load balancing config, isn't viable. The orgs that wins are the ones that give their developers the power, with the checks, to get to production without any, and i mean ZERO, in process approvals, rather the process itself is approved, and so long as you stay int he bounds you are good. what is an infra person but yet another person, there is no reason why these ip/fqdn decisions can't reside where they should, with the devs, with the ops teams providing the oversite.

additionally, tools like this shouldn't be imparting arbitrary restrictions based on what control models with hink clients want, rather, give them the options, then they can lock them down based on their policies. so, if you go with the annotation approach, or something like it, if client wants to lock that down, they can use other methods, such as admission controls, to prevent use. if the control is moved to a global CR that the devs have no chance of touch, we preclude those that desire to relinquish the control.

Reply:
@fedepaol I had an idea to perhaps meet the all requirements that we have.
What if we said that there is a controller that watches for  `metallb.io/assign-from-fqdn: myglobal.service.example.io` annotation. When this annotation is found the controller automatically creates an IPAddressPool like this:
```yaml
apiVersion: metallb.io/v1beta1
kind: IPAddressPool
metadata:
  name: example
  namespace: metallb-system
spec:
  sources:
      - type: dns
        fqdn: myglobal.service.example.io
```
And also assigns the service with that annotation pointing that pool to the Service: `metallb.universe.tf/address-pool: example`

This way we still the user to input the fqdn, and we use the IPAddressPool to assign the IP, like you wanted.

Reply:
I am more and more inclined to say this is something that should / can be handled by a controller that is external to MetalLB, as @oribon said when mentioning [external secrets](https://external-secrets.io/) with the mapping between the source (dns, dhcp) and the destination ipaddresspool defined in some CRD  (plus some property, like the service selector).

This would work and would also not add extra complexity to metallb, as it would be an external component. Given this, I think this must not be part of metallb but a sort of add-on. In the future, we might provide a reference one so users can write their own (and eventually contribute it), but this depends on our bandwith.

Also, this can be translated in your case with a controller that listens to the services, looks at the fqdn annotation to fetch (and reconciles) the ip.
When generating the ipaddresspool, this reconciler can add also a service selector so the addresspool is bound only to that service.